nexus/reconfigurator/planning/src/blueprint_builder/builder.rs

```diff
@@ -616,7 +675,187 @@ impl<'a> BlueprintBuilder<'a> {
             !removals.contains(&PhysicalDiskUuid::from_untyped_uuid(config.id))
         });
 
-        Ok(EnsureMultiple::Changed { added, removed })
+        Ok(EnsureMultiple::Changed { added, updated: 0, expunged: 0, removed })
+    }
+
+    /// Ensures that a sled in the blueprint has all the datasets it should.
+    ///
+    /// We perform the following process to decide what datasets should exist
+    /// in the blueprint during the planning phase:
+    ///
+    /// INPUT                    | OUTPUT
+    /// ----------------------------------------------------------------------
+    /// zpools in the blueprint  | blueprint datasets for debug, root filesystem
+    ///                          | (All zpools should have these datasets)
+    /// ----------------------------------------------------------------------
+    /// zones in the blueprint   | blueprint datasets for filesystems, durable data
+    ///                          | (These datasets are needed for zones)
+    /// ----------------------------------------------------------------------
+    /// discretionary datasets   | blueprint datasets for discretionary datasets
+    /// NOTE: These don't exist, |
+    /// at the moment            |
+    /// ----------------------------------------------------------------------
+    ///
+    /// From this process, we should be able to construct "all datasets that
+    /// should exist in the new blueprint".
+    ///
+    /// - If new datasets are proposed, they are added to the blueprint.
+    /// - If datasets are changed, they are updated in the blueprint.
+    /// - If datasets are not proposed, but they exist in the parent blueprint,
+    /// they are removed.
+    pub fn sled_ensure_datasets(
```

Comment:
This is the entrypoint for all our dataset planning. It's also the function from blueprint building I'm testing most directly down below.